### PR TITLE
gcd_of_n_numbers

### PR DESCRIPTION
**Pull request by balakhaniyan**
### Describe your change:

Greatest common divisor for n numbers

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: maths/gcd_of_n_numbers.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+def get_greatest_common_divisor(*numbers: int) -> int:
+    """
+    get gcd of n numbers:
+    >>> get_greatest_common_divisor(18, 45)
+    9
+    >>> get_greatest_common_divisor(23, 37)
+    1
+    >>> get_greatest_common_divisor(2520, 8350)
+    10
```
Comment (cclauss) [inline maths/gcd_of_n_numbers.py:56]: Tests for zero, negative numbers, floating point numbers, letters?

File: maths/gcd_of_n_numbers.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    if it is equal to zero, we check again with number // factor
+    else we increase factor by one
+    """
+
+    if not isinstance(number, int):
+        raise TypeError("number must be integer")
+
+    if factors is None:
+        # first call of get factors and so there are no founded factors
+        factors = Counter()
```
Comment (cclauss) [inline maths/gcd_of_n_numbers.py:31]: All falsey values are as dangerous as None.
```suggestion
    factors = factors or  Counter()
```

File: maths/gcd_of_n_numbers.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+    if number == factor:  # break condition
+        # all numbers are factors of itself
+        factors[factor] += 1
+        return factors
+
+    if number % factor > 0:
+        # if it is grater than zero
```
Comment (cclauss) [inline maths/gcd_of_n_numbers.py:39]:
```suggestion
        # if it is greater than zero
```

Comment (balakhaniyan) [reply maths/gcd_of_n_numbers.py]: Thank you, updated... please check and let me know if there are any problems

File: maths/gcd_of_n_numbers.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    Traceback (most recent call last):
+        ...
+    TypeError: number must be integer and greater than zero
+
+    factor can be all numbers from 2 to number that we check if number % factor == 0
+    if it is equal to zero, we check again with number // factor
+    else we increase factor by one
+    """
+
+    match number:
```
Comment (cclauss) [inline maths/gcd_of_n_numbers.py:38]: @dhruvmanila AlgorithemsKeeper is running on Py3.10, not 3.11.

Comment (balakhaniyan) [reply maths/gcd_of_n_numbers.py:38]: Must I change the code?

Comment (realstealthninja) [reply maths/gcd_of_n_numbers.py:38]: Yes

Comment (dhruvmanila) [reply maths/gcd_of_n_numbers.py:38]: This should be fixed.

**Comment by dhruvmanila** [comment]
@algorithms-keeper review

File: maths/gcd_of_n_numbers.py (hunk 4 above)
Comment (balakhaniyan) [reply maths/gcd_of_n_numbers.py:38]: @dhruvmanila can you help me that how can I fix it? It runs correctly on my computer

File: maths/gcd_of_n_numbers.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+    # we just need factors, not numbers itself
+    try:
+        data = [get_factors(number) for number in numbers]
+    except TypeError as e:
+        raise Exception("numbers must be integer and greater than zero") from e
+
+    same_factors: Counter = data[0]
+    for d in data[1:]:
+        same_factors = same_factors & d
```
Comment (cclauss) [inline maths/gcd_of_n_numbers.py:96]:
```suggestion
        same_factors &= d
```

File: maths/gcd_of_n_numbers.py [hunk 6]
```diff
@@ -87,23 +88,19 @@ def get_greatest_common_divisor(*numbers: int) -> int:
 
     # we just need factors, not numbers itself
     try:
-        data = [get_factors(number) for number in numbers]
+        same_factors, *factors = map(get_factors, numbers)
```
Comment (cclauss) [inline maths/gcd_of_n_numbers.py:91]: https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md#coding-style
> [List comprehensions and generators](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions) are preferred over the use of lambda, map, filter, reduce

https://docs.python.org/3/whatsnew/3.0.html#views-and-iterators-instead-of-lists

File: maths/gcd_of_n_numbers.py [hunk 7]
```diff
@@ -87,23 +88,19 @@ def get_greatest_common_divisor(*numbers: int) -> int:
...
         # `&` return common elements with smaller value (for Counter type)
 
     # now, same_factors is something like {2: 2, 3: 4} that means 2 * 2 * 3 * 3 * 3 * 3
-    mult = 1
     # power each factor and multiply
     # for {2: 2, 3: 4}, it is [4, 81] and then 324
-    for m in [factor ** same_factors[factor] for factor in same_factors]:
-        mult *= m
-    return mult
+    return reduce(lambda x, y: x * y, (factor ** power for factor, power in same_factors.items()))
```
Comment (cclauss) [inline maths/gcd_of_n_numbers.py]: This is going to be hard to understand and maintain.

**Review by cclauss — CHANGES_REQUESTED** [review]

**Review by cclauss — APPROVED** [review]
Thanks for doing this!!

**Pull request merged**